File: scripts/matlab_utils/create_trainval_test_sets.m
function create_trainval_test_sets(dset_dir, setlist, output_dir)

% create/check the output directory
check_output_dir(output_dir);

if strncmp(setlist.exp_kind, 'cat', 3)
    exp_id = false;
elseif strncmp(setlist.exp_kind, 'id', 2)
    exp_id = true;
else
    error('Unsupported exp_kind in the question config file!');
end

if isfield(setlist, 'divide_trainval_perc')
    divide_trainval_perc = setlist.divide_trainval_perc;
else
    divide_trainval_perc = false;
end

if divide_trainval_perc
    validation_perc = setlist.validation_perc;
    validation_split = setlist.validation_split; % 'step' or 'random'
end

% assign labels
fid_labels = fopen(fullfile(output_dir, 'labels.txt'), 'w');
if ~exp_id
    Y_digits = containers.Map (setlist.categories, 0:(length(setlist.categories)-1));
    for line=1:numel(setlist.categories)
        fprintf(fid_labels, '%s %d\n', setlist.categories{line}, Y_digits(setlist.categories{line}));
    end
else
    obj_names = repmat(setlist.categories, length(setlist.obj_per_cat.train),1);
    obj_names = obj_names(:);
    tmp = repmat(setlist.obj_per_cat.train', length(setlist.categories),1);
    obj_names = strcat(obj_names, tmp);
    
    Y_digits = containers.Map (obj_names, 0:(length(obj_names)-1));
    for line=1:length(obj_names)
        fprintf(fid_labels, '%s %d\n', obj_names{line}, Y_digits(obj_names{line}));
    end
end
fclose(fid_labels);


% create list of all files inside the 'iCW' dir
registry = registry_from_folder(dset_dir);
[registry_nofilename, ~, registry_ext] = cellfun(@fileparts, registry, 'UniformOutput', 0);

% select only images...
registry = registry(strcmp(registry_ext, '.jpg'));
registry_nofilename = registry_nofilename(strcmp(registry_ext, '.jpg'));


%% TRAINING set

% create desired folders...
categories = setlist.categories;
instances = setlist.obj_per_cat.train;
transformations = setlist.transf_per_obj.train;
days = setlist.day_per_transf.train;
if numel(days)==2
    days = {'day1','day2','day3','day4','day5','day6','day7','day8'};
elseif strcmp(days{1},'1')
    days = {'day1', 'day3', 'day5', 'day7'};
else
    days = {'day2', 'day4', 'day6', 'day8'};
end
cameras = {'left'};

selected_branches = allcomb(categories, instances, transformations, days, cameras);
selected_branches = [selected_branches(:,1) strcat(selected_branches(:,1), selected_branches(:,2)) selected_branches(:,3:end)];
selected_branches = fullfile(selected_branches(:,1), ...
    selected_branches(:,2), ...
    selected_branches(:,3), ...
    selected_branches(:,4), ...
    selected_branches(:,5));

% select only desired folders...
registry_train = registry(ismember(registry_nofilename, selected_branches));
registry_nofilename_train = registry_nofilename(ismember(registry_nofilename, selected_branches));

% get number of frames per folder...
[registry_dirs, ~, ic] = unique(registry_nofilename_train, 'stable'); % [C,ia,ic] = unique(A) % C = A(ia) % A = C(ic)
Ndirs = size(registry_dirs,1);
Nframes = zeros(Ndirs,1);
for ii=1:Ndirs
    Nframes(ii) = sum(ic==ii);
end
% ...and its range of indices in the Y arrays
startend = zeros(Ndirs+1,1);
startend(2:end) = cumsum(Nframes);

% eventually delete exceeding frames (randomly chosen)
if setlist.limit_frames.train
    tokeep_indices = zeros(setlist.max_frames.train, Ndirs);
    for ii=1:Ndirs
        if Nframes(ii)>=setlist.max_frames.train
            tokeep_indices(:, ii) = randperm(Nframes(ii), setlist.max_frames.train) + startend(ii);
        else
            tokeep_indices(:, ii) = 1:Nframes(ii) + startend(ii);
        end
    end
    registry_train = registry_train(tokeep_indices(:), :);
end

% eventually divide train and val
if divide_trainval_perc
    
    warning('Going to create validation set by splitting the training set...');
    
    if strcmp(validation_split, 'step')
        validation_step = 1/validation_perc;
        registry_val = registry_train(1:validation_step:end,:);
        registry_train(1:validation_step:end,:) = []; % train
    elseif strcmp(validation_split, 'random')
        fidxs = randperm(length(registry_train));
        split_idx = floor(validation_perc*length(registry_train));
        registry_val = registry_train(fidxs(1:split_idx),:);
        registry_train = registry_train(fidxs(split_idx+1:end),:);
    else
        error('Wrong validation_split in config file!');
    end
    
end

% assign Y
flist_splitted = regexp(registry_train, '/', 'split');
flist_splitted = vertcat(flist_splitted{:});
if exp_id
    Y = cell2mat(values(Y_digits, flist_splitted(:,2)));
else
    Y = cell2mat(values(Y_digits, flist_splitted(:,1)));
end

if divide_trainval_perc
    flist_splitted_val = regexp(registry_val, '/', 'split');
    flist_splitted_val = vertcat(flist_splitted_val{:});
    if exp_id
        Y_val = cell2mat(values(Y_digits, flist_splitted_val(:,2)));
    else
        Y_val = cell2mat(values(Y_digits, flist_splitted_val(:,1)));
    end
end

% write output
fid_Y = fopen(fullfile(output_dir, 'train.txt'), 'w');
if divide_trainval_perc
    fid_Y_val = fopen(fullfile(output_dir, 'val.txt'), 'w');
end

for line=1:numel(registry_train)
    fprintf(fid_Y, '%s %d\n', registry_train{line}, Y(line));
end
fclose(fid_Y);

if divide_trainval_perc
    for line=1:numel(registry_val)
        fprintf(fid_Y_val, '%s %d\n', registry_val{line}, Y_val(line));
    end
    fclose(fid_Y_val);
end

%% VALIDATION set

if ~divide_trainval_perc
    
    % create desired folders...
    categories = setlist.categories;
    instances = setlist.obj_per_cat.val;
    transformations = setlist.transf_per_obj.val;
    days = setlist.day_per_transf.val;
    if numel(days)==2
        days = {'day1','day2','day3','day4','day5','day6','day7','day8'};
    elseif strcmp(days{1},'1')
        days = {'day1', 'day3', 'day5', 'day7'};
    else
        days = {'day2', 'day4', 'day6', 'day8'};
    end
    cameras = {'left'};
    
    selected_branches = allcomb(categories, instances, transformations, days, cameras);
    selected_branches = [selected_branches(:,1) strcat(selected_branches(:,1), selected_branches(:,2)) selected_branches(:,3:end)];
    selected_branches = fullfile(selected_branches(:,1), ...
        selected_branches(:,2), ...
        selected_branches(:,3), ...
        selected_branches(:,4), ...
        selected_branches(:,5));
    
    % select only desired folders...
    registry_val = registry(ismember(registry_nofilename, selected_branches));
    registry_nofilename_val = registry_nofilename(ismember(registry_nofilename, selected_branches));
    
    % get number of frames per folder...
    [registry_dirs, ~, ic] = unique(registry_nofilename_val, 'stable'); % [C,ia,ic] = unique(A) % C = A(ia) % A = C(ic)
    Ndirs = size(registry_dirs,1);
    Nframes = zeros(Ndirs,1);
    for ii=1:Ndirs
        Nframes(ii) = sum(ic==ii);
    end
    % ...and its range of indices in the Y arrays
    startend = zeros(Ndirs+1,1);
    startend(2:end) = cumsum(Nframes);
    
    % eventually delete exceeding frames (randomly chosen)
    if setlist.limit_frames.val
        tokeep_indices = zeros(setlist.max_frames.val, Ndirs);
        for ii=1:Ndirs
            if Nframes(ii)>=setlist.max_frames.val
                tokeep_indices(:, ii) = randperm(Nframes(ii), setlist.max_frames.val) + startend(ii);
            else
                tokeep_indices(:, ii) = 1:Nframes(ii) + startend(ii);
            end
        end
        registry_val = registry_val(tokeep_indices(:), :);
    end
    
    % assign Y
    flist_splitted = regexp(registry_val, '/', 'split');
    flist_splitted = vertcat(flist_splitted{:});
    if exp_id
        Y = cell2mat(values(Y_digits, flist_splitted(:,2)));
    else
        Y = cell2mat(values(Y_digits, flist_splitted(:,1)));
    end
    
    % write output
    fid_Y = fopen(fullfile(output_dir, 'val.txt'), 'w');
    for line=1:numel(registry_val)
        fprintf(fid_Y, '%s %d\n', registry_val{line}, Y(line));
    end
    fclose(fid_Y);
    
end

%% TEST set


% create desired folders...
categories = setlist.categories;
instances = setlist.obj_per_cat.test;
transformations = setlist.transf_per_obj.test;
days = setlist.day_per_transf.test;
if numel(days)==2
    days = {'day1','day2','day3','day4','day5','day6','day7','day8'};
elseif strcmp(days{1},'1')
    days = {'day1', 'day3', 'day5', 'day7'};
else
    days = {'day2', 'day4', 'day6', 'day8'};
end
cameras = {'left'};

selected_branches = allcomb(categories, instances, transformations, days, cameras);
selected_branches = [selected_branches(:,1) strcat(selected_branches(:,1), selected_branches(:,2)) selected_branches(:,3:end)];
selected_branches = fullfile(selected_branches(:,1), ...
    selected_branches(:,2), ...
    selected_branches(:,3), ...
    selected_branches(:,4), ...
    selected_branches(:,5));

% select only desired folders...
registry_test = registry(ismember(registry_nofilename, selected_branches));
registry_nofilename_test = registry_nofilename(ismember(registry_nofilename, selected_branches));

% get number of frames per folder...
[registry_dirs, ~, ic] = unique(registry_nofilename_test, 'stable'); % [C,ia,ic] = unique(A) % C = A(ia) % A = C(ic)
Ndirs = size(registry_dirs,1);
Nframes = zeros(Ndirs,1);
for ii=1:Ndirs
    Nframes(ii) = sum(ic==ii);
end
% ...and its range of indices in the Y arrays
startend = zeros(Ndirs+1,1);
startend(2:end) = cumsum(Nframes);

% eventually delete exceeding frames
if setlist.limit_frames.test
    
    tokeep_indices = zeros(setlist.max_frames.test, Ndirs);
    
    for ii=1:Ndirs
        if Nframes(ii)>=setlist.max_frames.test
            
            if ischar(setlist.test_starting_img) && strcmp(setlist.test_starting_img, 'random')
                test_starting_img = ceil(1 + (Nframes(ii)-setlist.max_frames.test-1).*rand(1,1));
            elseif isnumeric(setlist.test_starting_img) && setlist.test_starting_img>0
                test_starting_img = min(Nframes(ii)-setlist.max_frames.test, setlist.test_starting_img);
            else
                error('Unsupported test_starting_img!')
            end
            tokeep_indices(:, ii) = ( test_starting_img:(test_starting_img+setlist.max_frames.test-1) ) + startend(ii);
        else
            tokeep_indices(:, ii) = 1:Nframes(ii) + startend(ii);
        end
    end
    
    registry_test = registry_test(tokeep_indices(:), :);
end

% assign Y
flist_splitted = regexp(registry_test, '/', 'split');
flist_splitted = vertcat(flist_splitted{:});
if exp_id
    Y = cell2mat(values(Y_digits, flist_splitted(:,2)));
else
    Y = cell2mat(values(Y_digits, flist_splitted(:,1)));
end

% write output
fid_Y = fopen(fullfile(output_dir, 'test.txt'), 'w');
for line=1:numel(registry_test)
    fprintf(fid_Y, '%s %d\n', registry_test{line}, Y(line));
end
fclose(fid_Y);

File: scripts/python_utils/create_train_val_test_sets.py
#!/usr/bin/python
import os
import itertools
from itertools import groupby
import numpy as np
import math


def create_train_val_test_sets(IMAGES_DIR, cfg, OUTPUT_DIR):

    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

    if cfg.EXP_KIND[:2] == 'ca':
        exp_id = False
    elif cfg.EXP_KIND[:2] == 'id':
        exp_id = True
    else:
        raise Exception('Unsupported exp_kind in the question config file!')

    if exp_id:
        lbl_idx = 1
    else:
        lbl_idx = 0

    if not exp_id:
        Ymap = dict([(val, idx) for idx, val in enumerate(cfg.CATEGORIES)])
        with open(os.path.join(OUTPUT_DIR, 'labels.txt'), 'w') as f:
            for val in cfg.CATEGORIES:
                f.write("%s %d\n" % (val, Ymap[val]))
    else:
        obj_names = [item for item in cfg.CATEGORIES for i in range(len(cfg.OBJ_PER_CAT.TRAIN))]
        tmp = [item for i in range(len(cfg.CATEGORIES)) for item in cfg.OBJ_PER_CAT.TRAIN ]
        obj_names = ["{}{}".format(a_, b_) for a_, b_ in zip(obj_names, tmp)]
        Ymap = dict([ (val, idx) for idx, val in enumerate(obj_names) ])
        with open(os.path.join(OUTPUT_DIR, 'labels.txt'), 'w') as f:
            for val in obj_names:
                f.write("%s %d\n" % (val, Ymap[val]))

    #registry = [val for sublist in [[os.path.relpath(os.path.join(i[0], j), IMAGES_DIR) for j in i[2]] for i in os.walk(IMAGES_DIR)] for val in sublist]
    #registry_nofilename = [os.path.dirname(item) for item in registry if os.path.splitext(item)[1] == '.jpg']
    #registry = [item for item in registry if os.path.splitext(item)[1] == '.jpg']

    registry = [os.path.join(root, name)
                 for root, dirs, files in os.walk(IMAGES_DIR)
                    for name in sorted([os.path.splitext(item)[0] for item in files if item.endswith(".jpg")], key=int)]
    registry_nofilename = [os.path.relpath(os.path.dirname(item), IMAGES_DIR) for item in registry]
    registry = [os.path.relpath(item, IMAGES_DIR)+".jpg" for item in registry]

    # TRAINING set

    categories = cfg.CATEGORIES
    instances = cfg.OBJ_PER_CAT.TRAIN
    transformations = cfg.TRANSF_PER_OBJ.TRAIN
    days = cfg.DAY_PER_TRANSF.TRAIN
    if len(days) == 2:
        days = ['day1', 'day2', 'day3', 'day4', 'day5', 'day6', 'day7', 'day8']
    elif days[0]=='1':
        days = ['day1', 'day3', 'day5', 'day7']
    else:
        days = ['day2', 'day4', 'day6', 'day8']
    cameras = ['left']

    selected_branches = itertools.product(categories, instances, transformations, days, cameras)
    selected_branches = list(selected_branches)
    selected_branches = [os.path.join(item[0], item[0]+item[1], os.path.join(*item[2:])) for item in selected_branches]

    registry_nofilename_train = [item for item in registry_nofilename if item in selected_branches]
    registry_train = [item for item in registry if os.path.dirname(item) in selected_branches]

    Nframes = [len(list(group)) for key, group in groupby(registry_nofilename_train)]
    Ndirs = len(Nframes)
    startend = [0] * (Ndirs + 1)
    startend[1:] = np.cumsum(Nframes)

    if cfg.LIMIT_FRAMES.TRAIN.ACTIVE:
        tokeep_indices = []
        for ii in xrange(0, Ndirs):
            tmp = np.random.permutation(Nframes[ii]) + startend[ii]
            tmp = tmp[:min(Nframes[ii], cfg.LIMIT_FRAMES.TRAIN.MAX)]
            tokeep_indices = np.concatenate((tokeep_indices, tmp), axis=0)
        tokeep_indices = tokeep_indices.astype(int)
        registry_train = [registry_train[i] for i in tokeep_indices]

    if cfg.DIVIDE_TRAINVAL_SET.ACTIVE:

        print 'Going to create validation set by splitting the training set...'

        if cfg.DIVIDE_TRAINVAL_SET.VALIDATION_SPLIT == 'step':
            validation_step = 1 / cfg.DIVIDE_TRAINVAL_SET.VALIDATION_PERC
            registry_val = registry_train[::int(validation_step)]
            del registry_train[::int(validation_step)]
        elif cfg.DIVIDE_TRAINVAL_SET.VALIDATION_SPLIT == 'random':
            fidxs = np.random.permutation(len(registry_train))
            fidxs = fidxs.astype(int)
            split_idx = int(math.floor(cfg.DIVIDE_TRAINVAL_SET.VALIDATION_PERC * len(registry_train)))
            registry_val = [registry_train[i] for i in fidxs[:split_idx]]
            registry_train = [registry_train[i] for i in fidxs[split_idx:]]
        else:
            raise Exception('Wrong validation_split in config file!')

    Ytrain = [Ymap[item.split(os.sep)[lbl_idx]] for item in registry_train]
    if cfg.DIVIDE_TRAINVAL_SET.ACTIVE:
        Yval = [Ymap[item.split(os.sep)[lbl_idx]] for item in registry_val]

    with open(os.path.join(OUTPUT_DIR, 'train.txt'), 'w') as f:
        for line in xrange(len(registry_train)):
            f.write("%s %d\n" % (registry_train[line], Ytrain[line]))

    if cfg.DIVIDE_TRAINVAL_SET.ACTIVE:
        with open(os.path.join(OUTPUT_DIR, 'val.txt'), 'w') as f:
            for line in xrange(len(registry_val)):
                f.write("%s %d\n" % (registry_val[line], Yval[line]))

    # VALIDATION set

    if not cfg.DIVIDE_TRAINVAL_SET.ACTIVE:

        categories = cfg.CATEGORIES
        instances = cfg.OBJ_PER_CAT.VAL
        transformations = cfg.TRANSF_PER_OBJ.VAL
        days = cfg.DAY_PER_TRANSF.VAL
        if len(days) == 2:
            days = ['day1', 'day2', 'day3', 'day4', 'day5', 'day6', 'day7', 'day8']
        elif days[0] == '1':
            days = ['day1', 'day3', 'day5', 'day7']
        else:
            days = ['day2', 'day4', 'day6', 'day8']
        cameras = ['left']

        selected_branches = itertools.product(categories, instances, transformations, days, cameras)
        selected_branches = list(selected_branches)
        selected_branches = [os.path.join(item[0], item[0]+item[1], os.path.join(*item[2:])) for item in selected_branches]

        registry_nofilename_val = [item for item in registry_nofilename if item in selected_branches]
        registry_val = [item for item in registry if os.path.dirname(item) in selected_branches]

        Nframes = [len(list(group)) for key, group in groupby(registry_nofilename_val)]
        Ndirs = len(Nframes)
        startend = [0] * (Ndirs + 1)
        startend[1:] = np.cumsum(Nframes)

        if cfg.LIMIT_FRAMES.VAL.ACTIVE:
            tokeep_indices = []
            for ii in xrange(0, Ndirs):
                tmp = np.random.permutation(Nframes[ii]) + startend[ii]
                tmp = tmp[:min(Nframes[ii], cfg.LIMIT_FRAMES.VAL.MAX)]
                tokeep_indices = np.concatenate((tokeep_indices, tmp), axis=0)
            tokeep_indices = tokeep_indices.astype(int)
            registry_val = [registry_val[i] for i in tokeep_indices]

        Yval = [Ymap[item.split(os.sep)[lbl_idx]] for item in registry_val]

        with open(os.path.join(OUTPUT_DIR, 'val.txt'), 'w') as f:
            for line in xrange(len(registry_val)):
                f.write("%s %d\n" % (registry_val[line], Yval[line]))

    # TEST set

    categories = cfg.CATEGORIES
    instances = cfg.OBJ_PER_CAT.TEST
    transformations = cfg.TRANSF_PER_OBJ.TEST
    days = cfg.DAY_PER_TRANSF.TEST
    if len(days) == 2:
        days = ['day1', 'day2', 'day3', 'day4', 'day5', 'day6', 'day7', 'day8']
    elif days[0] == '1':
        days = ['day1', 'day3', 'day5', 'day7']
    else:
        days = ['day2', 'day4', 'day6', 'day8']
    cameras = ['left']

    selected_branches = itertools.product(categories, instances, transformations, days, cameras)
    selected_branches = list(selected_branches)
    selected_branches = [os.path.join(item[0], item[0] + item[1], os.path.join(*item[2:])) for item in
                         selected_branches]

    registry_nofilename_test = [item for item in registry_nofilename if item in selected_branches]
    registry_test = [item for item in registry if os.path.dirname(item) in selected_branches]

    Nframes = [len(list(group)) for key, group in groupby(registry_nofilename_test)]
    Ndirs = len(Nframes)
    startend = [0] * (Ndirs + 1)
    startend[1:] = np.cumsum(Nframes)

    if cfg.LIMIT_FRAMES.TEST.ACTIVE:
        tokeep_indices = []
        for ii in xrange(0, Ndirs):
            if cfg.LIMIT_FRAMES.TEST.STARTING_IMG < 0:
                starting_idx = max(0, int(np.random.uniform(0, 1) * (Nframes[ii] - cfg.LIMIT_FRAMES.TEST.MAX)))
            else:
                starting_idx = max(0, min(cfg.LIMIT_FRAMES.TEST.STARTING_IMG, Nframes[ii] - cfg.LIMIT_FRAMES.TEST.MAX))
            tmp = range(starting_idx+ startend[ii], min(Nframes[ii]+ startend[ii], starting_idx + cfg.LIMIT_FRAMES.TEST.MAX+ startend[ii]))
            tokeep_indices = np.concatenate((tokeep_indices, tmp), axis=0)
        tokeep_indices = tokeep_indices.astype(int)
        registry_test = [registry_test[i] for i in tokeep_indices]

    Ytest = [Ymap[item.split(os.sep)[lbl_idx]] for item in registry_test]

    with open(os.path.join(OUTPUT_DIR, 'test.txt'), 'w') as f:
        for line in xrange(len(registry_test)):
            f.write("%s %d\n" % (registry_test[line], Ytest[line]))
